File: Contest/Contest_34/D_Jzzhu_and_Cities.cpp
#include <bits/stdc++.h>

using namespace std;

struct Edge
{
  int dest;
  int length;
  bool isTrain;
};


int main() {
    int n, m, k;
    cin >> n >> m >> k;

    const int INF = numeric_limits<int>::max();
    // make graph vector of of Edge struct length n+1
    vector<Edge> graph(n+1);


    for (int i = 0; i < m; i++) {
        int src, dest, length;
        cin >> src >> dest >> length;
        
    }

    for (int i = 0; i < k; i++) {
        int dest, length;
        cin >> dest >> length;
        // assign graph[1] to an edge with dest and length and False
        
        graph[1].push_back(make_pair(dest, length));
    }

    map<int, int> distances;
    for (int i = 1; i <= n; i++) {
        distances[i] = INF;
    }
    distances[1] = 0;

    set<int> visited;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push(make_pair(0, 1));

    while (!pq.empty()) {
        int curr_dis = pq.top().first;
        int curr_city = pq.top().second;
        pq.pop();

        if (visited.find(curr_city) != visited.end()) {
            continue;
        }
        visited.insert(curr_city);

        for (auto neighbor : graph[curr_city]) {
            int nbr = neighbor.first;
            int dis = neighbor.second;
            int new_dis = curr_dis + dis;

            if (new_dis < distances[nbr]) {
                distances[nbr] = new_dis;
                pq.push(make_pair(new_dis, nbr));
                k--;
            }
        }
    }

    cout << k << endl;

    return 0;
}

File: Contest/Contest_34/D_Jzzhu_and_Cities.py
# from collections import defaultdict
# from heapq import heappush, heappop

# n, m, k = map(int, input().split())
# INF = float('inf')
# graph = defaultdict(list)
# for _ in range(m):
#     src, dest, length = map(int, input().split())
#     graph[src].append((dest, length, False))
#     graph[dest].append((src, length, False))

# for _ in range(k):
#     dest, length = map(int, input().split())
#     graph[1].append((dest, length, True))
#     graph[dest].append((1, length, True))

# distances = defaultdict(lambda: INF)
# distances[1] = 0
# visited = set()

# heap = [(0, 1, False)]
# while heap:
#     curr_dis, curr_city, was_train = heappop(heap)
#     if curr_city in visited:
#         continue
#     visited.add(curr_city)

#     for nbr, dis, is_train in graph[curr_city]:
#         new_dis = curr_dis + dis
#         if curr_dis != INF and new_dis < distances[nbr]:
#             distances[nbr] = new_dis
#             if is_train and not was_train:
#                 heappush(heap, (new_dis, nbr, True))
#                 k -= 1
#             if was_train and not is_train:
#                 heappush(heap, (new_dis, nbr, False))
#                 k += 1
#         elif curr_dis != INF and new_dis == distances[nbr]:
#             if was_train and not is_train:
#                 k += 1
# print(k)



import heapq

def dijkstra(graph, start):
    n = len(graph)
    dist = [float('inf')] * n
    dist[start] = 0
    min_heap = [(0, start)]

    while min_heap:
        d, u = heapq.heappop(min_heap)
        if d > dist[u]:
            continue

        for v, w in graph[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                heapq.heappush(min_heap, (dist[v], v))

    return dist

def main():
    n, m, k = map(int, input().split())
    graph = [[] for _ in range(n)]

    for _ in range(m):
        u, v, x = map(int, input().split())
        u -= 1
        v -= 1
        graph[u].append((v, x))
        graph[v].append((u, x))

    train_routes = []
    for _ in range(k):
        s, y = map(int, input().split())
        s -= 1
        train_routes.append((s, y))

    capital_distances = dijkstra(graph, 0)

    max_closed_routes = 0
    for s, y in train_routes:
        if capital_distances[s] == y:
            max_closed_routes += 1

    print(max_closed_routes)

if __name__ == "__main__":
    main()
